File: scripts/build_excuses_js.py
#!/usr/bin/env python3
# Reads scripts/excuses.csv and writes assets/js/excuses.js as an ES module.
# Usage: python scripts/build_excuses_js.py
# Optional: python scripts/build_excuses_js.py --input other.csv --output assets/js/excuses.js

import csv, argparse, pathlib, sys

def main():
    parser = argparse.ArgumentParser()
    parser.add_argument("--input", "-i", default="scripts/excuses.csv")
    parser.add_argument("--output", "-o", default="assets/js/excuses.js")
    args = parser.parse_args()

    in_path = pathlib.Path(args.input)
    out_path = pathlib.Path(args.output)
    if not in_path.exists():
        print(f"Input CSV not found: {in_path}", file=sys.stderr)
        sys.exit(1)

    rows = []
    with in_path.open(newline="", encoding="utf-8") as f:
        reader = csv.DictReader(f)
        for r in reader:
            r["id"] = int(r["id"])
            r["category"] = r["category"].strip()
            r["intent"] = r["intent"].strip()
            r["tone"] = r["tone"].strip().lower()
            r["text"] = r["text"].strip()
            rows.append(r)

    rows.sort(key=lambda x: x["id"])

    allowed_tones = {"formal","casual"}
    for r in rows:
        if r["tone"] not in allowed_tones:
            print(f"Warning: row id={r['id']} has non-standard tone '{r['tone']}'. Allowed: {allowed_tones}", file=sys.stderr)

    def js_escape(s: str) -> str:
        return s.replace("\\", "\\\\").replace("`", "\\`")

    parts = []
    for r in rows:
        parts.append(
            "  { id: %d, category: '%s', intent: '%s', tone: '%s', text: `%s` }"
            % (r["id"], r["category"].replace("'", "\\'"), r["intent"].replace("'", "\\'"),
               r["tone"], js_escape(r["text"]))
        )

    content = "export const EXCUSES = [\\n" + ",\\n".join(parts) + "\\n];\\n"

    out_path.parent.mkdir(parents=True, exist_ok=True)
    out_path.write_text(content, encoding="utf-8")
    print(f"Wrote {out_path} with {len(rows)} excuses.")

if __name__ == "__main__":
    main()

File: scripts/build_excuses_js.mjs
// Usage:
//   node scripts/build_excuses_js.mjs [inputCSV] [outputJS]
// Defaults to scripts/excuses.csv -> assets/js/excuses.js

import { readFile, writeFile, mkdir } from 'node:fs/promises';
import { dirname } from 'node:path';

const input = process.argv[2] || 'scripts/excuses.csv';
const output = process.argv[3] || 'assets/js/excuses.js';

function parseCSV(text){
  // Minimal CSV parser supporting quoted fields with commas and quotes "".
  const rows = [];
  let i=0, field='', row=[], inQuotes=false;
  while (i <= text.length){
    const c = text[i] ?? '\n';
    if (inQuotes){
      if (c === '"'){
        if (text[i+1] === '"'){ field += '"'; i++; }
        else inQuotes = false;
      } else { field += c; }
    } else {
      if (c === '"'){ inQuotes = true; }
      else if (c === ','){ row.push(field); field=''; }
      else if (c === '\n' || c === '\r'){
        if (field !== '' || row.length){ row.push(field); rows.push(row); row=[]; field=''; }
        if (c === '\r' && text[i+1] === '\n') i++;
      } else { field += c; }
    }
    i++;
  }
  return rows;
}

const raw = await readFile(input, 'utf8');
const rows = parseCSV(raw).filter(r => r.length);
const header = rows.shift().map(h => h.trim());
const idx = Object.fromEntries(header.map((h, i) => [h, i]));

const data = rows.map(r => ({
  id: Number(r[idx.id]),
  category: r[idx.category].trim(),
  intent: r[idx.intent].trim(),
  tone: (r[idx.tone] || '').trim().toLowerCase(),
  text: (r[idx.text] || '').trim(),
})).sort((a,b)=>a.id-b.id);

const js = 'export const EXCUSES = [\\n' + data.map(e => {
  const esc = e.text.replaceAll('\\\\', '\\\\\\\\').replaceAll('`','\\\\`');
  return `  { id: ${e.id}, category: '${e.category.replaceAll(\"'\",\"\\\\'\")}', intent: '${e.intent.replaceAll(\"'\",\"\\\\'\")}', tone: '${e.tone}', text: \\`${esc}\\` }`;
}).join(',\\n') + '\\n];\\n';

await mkdir(dirname(output), { recursive: true });
await writeFile(output, js, 'utf8');
console.log(`Wrote ${output} with ${data.length} excuses.`);
